apis/Google.Cloud.Datastore.V1/Google.Cloud.Datastore.V1/DatastoreDbBuilder.cs

```diff
@@ -140,29 +85,32 @@ void ApplyEmulatorConfiguration()
         protected override GrpcAdapter DefaultGrpcAdapter =>
             throw new InvalidOperationException($"This property should never execute in {nameof(DatastoreDbBuilder)}");
 
-        private class EmulatorConfiguration
+        /// <summary>
+        /// Common code for handling project ID, namespace ID and emulator detection.
+        /// </summary>
+        private ConfiguredBuilder PrepareBuilder()
         {
-            internal const string EmulatorHostVariable = "DATASTORE_EMULATOR_HOST";
-            internal const string EmulatorProjectVariable = "DATASTORE_PROJECT_ID";
-            internal string Endpoint { get; }
-            internal string ProjectId { get; }
-
-            private EmulatorConfiguration(string endpoint, string projectId)
+            var environment = GetEmulatorEnvironment(s_requiredEmulatorVariables, s_allEmulatorVariables);
+            if (environment is object)
             {
-                Endpoint = endpoint;
-                ProjectId = projectId;
+                // Note: this is a recursive call, but because EmulatorDetection defaults to None,
+                // it will only recurse once.
+                return new DatastoreDbBuilder
+                {
+                    Endpoint = environment[s_emulatorHostVariable],
+                    Settings = Settings,
+                    ProjectId = ProjectId ?? environment[s_emulatorProjectVariable],
```

Comment:
I know it was like this before, but intuitively I would have though the emulator project ID would have precedence when using the emulator. Flagging just in case.

Reply:
I *think* it makes more sense to make code take precedence here, but I do see your point. Let's wait and see if anyone flags it :)